### ISSUE TITLE
Media Library crashes with an error message while loading one or more images

### ISSUE THREAD

**Issue by pgsfredda** [strapi/strapi#18175]
### Required System information
- Strapi Version: 4.13.7
- Operating System: Ubuntu 20.04.6 LTS
- Database: mySQL 8.0.34 (using mysql2 as driver)
- Node Version: 18.18.0
- NPM Version: 9.8.1
- Yarn Version: none
- Dev language: typescript

### Describe the bug
When I try to load one or more images (size: 5kb) Strapi Admin get an error message:
`Woops! Something went wrong. Please, try again.`

### Steps to reproduce the behavior

1. Go to `Media Library`
2. Click on `Add new assets`
3. Browsing files or dragging the image (max image size used is 5kb)
4. See error `Woops! Something went wrong. Please, try again.`
5. The error reported in the console is:

![immagine](https://github.com/strapi/strapi/assets/7228819/5d3e8220-8390-4b45-a790-745c7966b4ae)

No error messages on server log.

### Additional context

The production strapi configuration is under the sub path `/strapi`

My `config/server.ts`:

~~~js
export default ({ env }) => ({
  host: env('HOST', '0.0.0.0'),
  port: env.int('PORT', 1337),
  url: env('STRAPI_URL', '/strapi'),
  app: {
    keys: env.array('APP_KEYS'),
  },
  webhooks: {
    populateRelations: env.bool('WEBHOOKS_POPULATE_RELATIONS', false),
  },
});
~~~

My `config/admin.ts`:
~~~js
export default ({ env }) => ({
  url: env('STRAPI_ADMIN_URL', '/strapi/admin'),
  auth: {
    secret: env('ADMIN_JWT_SECRET'),
  },
  apiToken: {
    salt: env('API_TOKEN_SALT'),
  },
  transfer: {
    token: {
      salt: env('TRANSFER_TOKEN_SALT'),
    },
  },
});
~~~

No `api prefix` configured in `config/admin.ts`.

Any other configuration do not work with my apache proxy configuration (`HTTP 404 Not found` error message emitted by strapi).

Admin build command:
`NODE_ENV=production STRAPI_URL=/strapi STRAPI_ADMIN_URL=/strapi/admin npm run build`

My apache proxy config:
~~~code
<IfModule mod_ssl.c>
<VirtualHost *:443>
        ServerName mydomain
        ServerAlias mydomain
        ServerAdmin myemail
        DocumentRoot /var/www/mydomain/html

        <Directory /var/www/mydomain/html>
            Options Indexes FollowSymLinks
            AllowOverride All
            Require all granted
        </Directory>

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined

        <IfModule mod_dir.c>
            DirectoryIndex index.php index.pl index.cgi index.html index.xhtml index.htm
        </IfModule>

	ProxyPreserveHost On
  	ProxyRequests Off
	ProxyPass /strapi http://localhost:1337
	ProxyPassReverse /strapi http://localhost:1337
</VirtualHost>
</IfModule>
~~~

When I make the same test on local machine, without server url field definition, strapi admin works fine.

Can someone help me? Thank you in advance

**Comment by pgsfredda** [comment]
# UPDATE

I made some tests with the file `appendSearchParamsToUrl.js` adding some rows to modify the `url` and `window.strapi.backendURL` parameters.
The solution is not good because the problem is the composition of those parameters  elsewhere but with this change I can use media now. 

I'm waiting for a correct patch. **This is a blocking issue**.
Thank you

### Temporary patch for `appendSearchParamsToUrl.js`
~~~js
/**
 * Append the given search params to the given URL.
 *
 * @param {String} url The URL string to append the search params to
 * @param {Object} params The object of search params to append to the URL
 * @returns {String} A string representing the URL with the search params appended
 */
const appendSearchParamsToUrl = ({ url, params }) => {
  if (url === undefined || typeof params !== 'object') {
    return url;
  }

  var blob = url.split(':')[0] === 'blob'? 'blob:': '';
  if(blob) url = url.replace('blob:', '');
  url = url.replace(window.origin, '').replace(window.strapi.backendURL, '');
  if(!window.strapi.backendURL.startsWith(window.origin)) window.strapi.backendURL = window.origin + window.strapi.backendURL;
  if(!url.startsWith(window.strapi.backendURL)) url = blob + window.strapi.backendURL + url;
  
  console.log('*** STRAPI *** append', {url, params, origin: window.origin, backendURL: window.strapi.backendURL});
  const urlObj = new URL(url, window.strapi.backendURL);

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined) {
      urlObj.searchParams.append(key, value);
    }
  });

  return urlObj.toString();
};

export { appendSearchParamsToUrl };

~~~

**Comment by Boegie19** [comment]
@joshuaellis Mind doing the labels for this one on Monday since I don't know what you have already done with this?

**Comment by DonikaV** [comment]
Hey, i am on 4.14.2 and have the same issue. How can I fix it?

**Comment by pgsfredda** [comment]
Hi.
If you have the same issue with media library you can try to put the file here:
`.../node_modules/@strapi/plugin-upload/admin/src`

But I'm facing again with the problem also with this mini mini patches.

The problem is elsewhere.

@joshuaellis, thank you for your work; strapi is a great product but if you use the Media library the `severity`is not only `medium`. It is impossible the use of the Media library at the moment.

Thank you

**Comment by Boegie19** [comment]
@pgsfredda For you as it would be a critical. but since strapi needs to look at the bigger scale and only 2 people are confirmed having this issue. upgrading it to a high would mean it would be a much bigger group of users impact. 

still this is how I see it from the strapi community side and I can't talk for strapi.

**Comment by segovia** [comment]
@Boegie19 I also confirm to have this issue but my ticket was closed in favor of this one, even though I opened mine earlier. In any case, I consider this issue as severe. The documentation says relative path urls are allowed but this bug disables a crucial part of the application. Together with the fact that the fix should be very small, I really can't see why this is not fixed sooner rather than later.

**Comment by joshuaellis** [comment]
@pgsfredda @segovia if either of you think the issue is more important for yourselves then we accept PRs. However, @Boegie19 is correct – our internal escalation policies don't regard this as higher than medium. In fact, in reality it should be marked as low because the workaround is to use an absolute URL which is what we recommend anyway.

**Comment by pgsfredda** [comment]
@joshuaellis I’m sorry for my English which is not so fluent and thank you for your work. I love Strapi.

But, as you can see in my description of the long problem, I am not using a relative path, but a secondary /strapi path starting from the root.
Another problem is, as you can see in my description, the configuration of apache proxy that does not work if I use another solution (I suggest to add in the docs area an explanation of configuration specific to apache in addition to ngix).

In fact, with this configuration, to reach my admin app, I have to insert twice strapi: `https://mydomain/strapi/strapi/strapi/admin`

Maybe (but I don’t know strapi so well) the problem is the url composition in another function. My simple patch (which now no longer works) cleans the path and add some tokens to complete correcly the url.

More generally, the Media Library is a strapi admin feature that now, under my configuration, does not work. This is why I wrote that gravity is not medium. Maybe, you can write a new configuration document or fix the problem. I don't know wich is the better solution.

But I can wait for few time.

Again, thanks for your work. Strapi is great. I’m using it in a very important project.

Best regards

**Comment by schlotterich** [comment]
I'm running into the same problem after updating to Strapi Version 15.0.2. The issue does not occur in the local development environment but persists in both staging and production. When I reverted to Version 4.12.4, the problem disappeared.

**Comment by moonstar-x** [comment]
@matheusdamiao Hey, if it is of any help:

I am currently in 4.13.3 and I managed to fix it right now on my end with the following:

Inside the `config/server.js` file, make sure you have a `url` property that resolves to the URL of your Strapi service, (i.e, `http://localhost:1337`, you could use an env variable too.

I then removed the `.cache`, `.tmp` and `build` folders from my Strapi project (this deletes your data so make sure you have an exported backup or something, you would still need to regenerate your api tokens if you're using them) and re-built it with `strapi build` and imported my data with `strapi import` and it was working.

---

To be fair, you could probably try removing only the `.cache` folder and it _could_ work, but I wanted to retry everything with a clean version.

**Comment by CubedEye** [comment]
This [PR](https://github.com/strapi/strapi/pull/18028), or this [issue](https://github.com/strapi/strapi/issues/18024)? doesn't take into account the case where you have a relative `server.url` in the server config at build time, so that propagates through to `STRAPI_ADMIN_BACKEND_URL`, which  is set but not to an absolute URL. This propagates through to the client where `window.strapi.backendURL` is set to the relative URL that was set.

<img width="837" alt="Screenshot 2023-12-05 at 11 53 02 am" src="https://github.com/strapi/strapi/assets/993865/46882d01-e11f-4d81-b6f2-093ec51f8821">

Per the docs, this [url](https://docs.strapi.io/dev-docs/configurations/server) can be relative even though it says it's recommended to be absolute. This behaviour used to work in 4.12.7 and below. However since this change, the media library has been broken for everyone who is running Strapi at a relative URL to their domain.

This only seems to break the upload plugin because of the changes to `packages/core/upload/admin/src/utils/appendSearchParamsToUrl.js` which is trying to use that backendURL as the origin base for `new URL`. I'm not sure if this is all the places this has changed since 4.13 through to 4.15. But 4.12.7 also have the relative `window.strapi.backendURL` and works fine.

**Comment by jameshhood** [comment]
> In fact, with this configuration, to reach my admin app, I have to insert twice strapi: `https://mydomain/strapi/strapi/admin`

In your config/admin.js file, change your `url: env(..., '/strapi/admin')` to just '/admin' and that should fix that

On a side note:

> In fact, in reality it should be marked as low because the workaround is to use an absolute URL which is what we recommend anyway.

We define the host, the port, why does it need to be absolute? Everything else works relative. This is super critical for me because I have an environment which cannot go live because we cannot manage the media and make sure things are kosher

**Comment by jameshhood** [comment]
According to this strapi [forum](https://forum.strapi.io/t/media-library-url-error/31861/8) post the issue comes from this [commit](https://github.com/strapi/strapi/commit/4d92b11512b1ac226c305ec14aea26d6726fb795#diff-de54f7f14b12a5944c35221d996fd0db5817230ef0b0139974dbab112f13fe78).

So by patching the appendSearchParamsToUrl.js file then it should fix this. I can't exactly do this though because I have an automated build process that cleans and ensures a fresh environment each time so I can't exactly just edit that during that process. Please make this a critical issue. I've seen numerous people posting about this around in the forums.

**Comment by cjambrosi** [comment]
Any predictions for fixing this problem?

**Comment by TripShade** [comment]
God grant us forgiveness for the horrors we are inflicted upon the world to get around this.

**Comment by CubedEye** [comment]
> us

@TripShade what have you done to get around this? 

We haven't upgraded since it broke at 4.12.7, I'm just about ready to patch-package this change back out, seeing there hasn't been any response from Strapi here or in the forum where I posted about this. Either they're not going to fix this or they don't know how, but some communication would be nice!

**Comment by TripShade** [comment]
```
[INFO] Including the following ENV variables as part of the JS bundle:
    - ADMIN_PATH
    - STRAPI_ADMIN_BACKEND_URL
    - STRAPI_TELEMETRY_DISABLED
```

was the last puzzle piece I needed. 

Instead of:
```ts
//server.ts
export default ({ env }) => ({
  host: env("HOST", "0.0.0.0"),
  port: env.int("PORT", 1337),
  url: env("PUBLIC_URL", "/strapi"),
  proxy: env.bool("PROXY", true),
  app: {
    keys: env.array("APP_KEYS"),
  },
  webhooks: {
    populateRelations: env.bool("WEBHOOKS_POPULATE_RELATIONS", false),
  },
});

//admin.ts
export default ({ env }) => ({
  auth: {
    secret: env("ADMIN_JWT_SECRET"),
  },
  apiToken: {
    salt: env("API_TOKEN_SALT"),
  },
  transfer: {
    token: {
      salt: env("TRANSFER_TOKEN_SALT"),
    },
  },
  flags: {
    nps: env.bool("FLAG_NPS", true),
    promoteEE: env.bool("FLAG_PROMOTE_EE", true),
  },
});
```

use:
```
//server.ts
export default ({ env }) => ({
  host: env("HOST", "0.0.0.0"),
  port: env.int("PORT", 1337),
  url: env("STRAPI_ADMIN_BACKEND_URL", "https://#somedomain#/strapi"),
  proxy: env.bool("PROXY", true),
  app: {
    keys: env.array("APP_KEYS"),
  },
  webhooks: {
    populateRelations: env.bool("WEBHOOKS_POPULATE_RELATIONS", false),
  },
});

//admin.ts
export default ({ env }) => ({
  auth: {
    secret: env("ADMIN_JWT_SECRET"),
  },
  apiToken: {
    salt: env("API_TOKEN_SALT"),
  },
  transfer: {
    token: {
      salt: env("TRANSFER_TOKEN_SALT"),
    },
  },
  flags: {
    nps: env.bool("FLAG_NPS", true),
    promoteEE: env.bool("FLAG_PROMOTE_EE", true),
  },
  url: env("ADMIN_PATH", "/dashboard"),
});
```

There is just no way getting around the absolute path. You can build without absolute path and run container with absolute path and most things wil work except uploads. Using the hinted ENV names fixes uploads for me.

Also since this might help someone else in the trences.
```yaml
ingress:
  enabled: true
  className: "nginx"
  annotations:
    nginx.ingress.kubernetes.io/backend-path-prefix: /
    nginx.ingress.kubernetes.io/ssl-redirect: "true"
    nginx.ingress.kubernetes.io/use-regex: "true"
    nginx.ingress.kubernetes.io/configuration-snippet: |
      rewrite ^(/strapi)(/|$)(.*) $2$3 break;
  hosts:
    - host: #somedomain#
      paths:
        - path: /
          pathType: Prefix
          backend:
            service:
              name: #someproject#-frontend-web
              port:
                number: 8080
        - path: /strapi(/|$)(.*)
          pathType: ImplementationSpecific
          backend:
            service:
              name: #someproject#-backend
              port:
                number: 1337
```

But to answer your question I had some nasty post build "RUN sed" statements towards /dist/build/main*js in my container script.

**Comment by segovia** [comment]
In the meantime, our project is is patching the strapi lib (we are currently using 4.19.0):

```
diff --git a/node_modules/@strapi/plugin-upload/dist/_chunks/index-erWFC6Q2.mjs b/node_modules/@strapi/plugin-upload/dist/_chunks/index-erWFC6Q2.mjs
index 7c994a742643f30a3a751e89393467c660b315f3..3b577b3cf7522ebed133122164b4f90a605ed36a 100644
--- a/node_modules/@strapi/plugin-upload/dist/_chunks/index-erWFC6Q2.mjs
+++ b/node_modules/@strapi/plugin-upload/dist/_chunks/index-erWFC6Q2.mjs
@@ -151,7 +151,8 @@ const appendSearchParamsToUrl = ({ url, params }) => {
   if (url === void 0 || typeof params !== "object") {
     return url;
   }
-  const urlObj = new URL(url, window.strapi.backendURL);
+  // FIX FOR https://github.com/strapi/strapi/issues/18175
+  const urlObj = new URL(url, window.location.origin);
   Object.entries(params).forEach(([key, value]) => {
     if (value !== void 0) {
       urlObj.searchParams.append(key, value);
```

**Comment by CubedEye** [comment]
@segovia I figured someone would have patched already, thanks for posting

**Comment by cjambrosi** [comment]
@TripShade Could you put the values of the environment variables as an example?

**Comment by jameshhood** [comment]
I just don't understand why they are releasing a version 5 beta where they explicitly say "Do not use this for production" When there are clear production problems here in version 4. 

I had another random bug I came across from reverting back to an old version to fix this bug and now I'm full circle. I use devops to build my source so having custom packages is not really something I can use as a work around. I can however define the environment variable after it's been deployed to our k8s...except that the environment variable is only really used during build time.

I understand the recommendation for an absolute path however due to the `window.strapi.backendURL` being something built into the app and not something referenced at runtime... the whole medium severity bit to this is a bit rubbish for having a work around... Maybe in a development environment but you can't honestly expect us to deploy this in dev right? Plenty of people have mentioned the fix. Can this not just be fixed in the project?

I literally just scaffold up a new project running 4.21 and this doesn't work...a CMS that doesn't allow you to manage media content...

**Comment by CubedEye** [comment]
@joshuaellis It looks like this has been assigned to you and you profile suggests you are working on Strapi, can we get some communication from Strapi about this? There is clearly a problem here and the silence from Strapi is helping anyone.

**Comment by CubedEye** [comment]
> @jameshhood I understand the recommendation for an absolute path however due to the window.strapi.backendURL being something built into the app and not something referenced at runtime... the whole medium severity bit to this is a bit rubbish for having a work around... Maybe in a development environment but you can't honestly expect us to deploy this in dev right? Plenty of people have mentioned the fix. Can this not just be fixed in the project?

Our problem is we build this once on CI and deploy to 3 different environments (testing, staging and prod). So either I now need to build this 3 times with 3 different URLs? Seems like a bit of a cop-out, when this could be configured at runtime and this would work like the rest of Strapi.

Anyway, we're going to patch this in the meantime, until someone offers something better

**Comment by jameshhood** [comment]
Yeah same. We essentially host strapi instances for multiple apps which have multiple environments. We check our code in and then have a pipeline which builds the app automatically and publishes a docker image to aws to then pull and run in our cluster for each app/environment. 

I think for now we are just going to stop building the app in the pipeline and storing the output in the docker image to now just store the whole source on there which is stupid; but every time our docker image fires up, it's going to build the app with the provided environment variable and then run it. That way the environment variable propagates through to the build output properly. Not even sure that's going to work.

This bug has been floating around now for 6 months now which has been blocking numerous people from even updating strapi. That alone should be encouragement to fix this.

**Comment by joshuaellis** [comment]
Hey @CubedEye, referring to this comment:

> doesn't take into account the case where you have a relative server.url in the server config at build time, so that propagates through to STRAPI_ADMIN_BACKEND_URL, which is set but not to an absolute URL. This propagates through to the client where window.strapi.backendURL is set to the relative URL that was set.

and looking at the screenshot you shared (v useful) your backend url is `/cms` so could you therefore assume that if your admin panel was hosted on `https://my-cms.com/admin` your backend URL (if it were to be absolute) would be `https://my-cms.com/cms`? In which case, we could potentially support the use of relative URLs by taking the `window.location.origin` and appending the `BACKEND_URL` to it. Or would your use of relative url be something different? 😄 

And just so I understand and not in any way provocative in your circumstance why are you using relative over absolute? :)

**Comment by CubedEye** [comment]
@joshuaellis, we're currently hosting this as a subpath of our main domain and proxy the requests through to the container inside AWS, ie.

`http://domain.com/strapi` with the admin at `http://domain.com/strapi/admin`, but we also have it running on our testing server `http://test.domain.com/strapi/admin`. But we proxy this through to Strapi running underneath to something like you've suggested, except it's not public - `http://internal-ip/admin`

We currently build Strapi once on our CI environment (single docker image) with`STRAPI_ADMIN_BACKEND_URL` set to `""` so it configures everything relative, but then run it live with that set to the environment that it's running on (ie. testing, staging, production).  This is already hack around needing to define it at build time, but seems to work for everything else. Different problem to this specific issue, but would be nice to have this configurable at runtime for all of Strapi, so it can support single build, multiple environments.

As for this issue in particular, using `window.location.origin` if `BACKEND_URL` isn't set as you've suggested would probably solve this issue in the simplest way, however I'm not sure what knock on effects this would have to the rest of Strapi (none I don't think, because nothing else seems to be broken). This would keep the security intent of the code that broke this, but also make it work again 🤞

**Comment by joshuaellis** [comment]
RE. using `process.env` at runtime isn't really doable because it doesn't exist in the browser and strapi is an SPA, so i'm not sure it's viable at the moment. But duly noted that it would be helpful :)

Would you be interested in trialing an experimental version of strapi with this fix to validate it solves your issue? If it doesn't then we can investigate and discuss further of course!

**Comment by CubedEye** [comment]
@joshuaellis happy help however we can.

**Comment by joshuaellis** [comment]
experimental version containing what i hope is the fix – 0.0.0-experimental.bb0b6c425a787a0a8268bb133787d71713c56944 

NOTE! you may get NPM warnings about peer deps, this is expected because it's an experimental version.

**Comment by CubedEye** [comment]
@joshuaellis looks like this change got merged already... 🚀 Thank you!  

I'm just confirming that it did indeed fix our specific issue. However, I am also doing some digging because it looks like my issue was fixed for us sometime after 4.17.0, but I'm still confirming that is the case and I didn't have cached docker containers getting in the way.
